Add timestamp to solution filenames to prevent overwrite

girish-kanjiyani7 opened the pull request:
This PR introduces an enhancement to allow multiple solutions for the same LeetCode problem to be saved uniquely in the GitHub repo, avoiding overwrites and preserving version history.

**_Key changes:_**
Modified the loader() function to append a timestamp to each solution filename.

Utilized the existing getTodaysDate() function to generate a clean, filesystem-safe timestamp.

Format used: MM-DD-YYYY_HH-MM-SS

Example filename: 0001-two-sum-03-25-2025_15-30-12.py

If a right-click suffix is provided, it is combined with the timestamp (e.g., 0001-two-sum-dfs-03-25-2025_15-30-12.cpp).

**_Why This Is Useful:_**
Maintains a complete history of all attempts/strategies for the same problem.

Helps track progress and improvement over time.

Prevents accidental overwrites when using the "Push" button or auto-upload after resubmission.

 **_Additional Notes:_**
Changes are backward-compatible. Default behavior (with overwrite) is preserved if users modify the logic back.

Timestamp-based filenames are especially useful for practicing multiple approaches (e.g., brute force, optimized, recursive, etc.).


Thanks for maintaining this amazing project — it’s been incredibly helpful. I hope this feature can add value to the LeetHub user base!

raphaelheinz commented:
Hi @girish-kanjiyani7, many thanks for your PR :)

raphaelheinz commented:
I really appreciate your PR 🙏 Already checked and everything is fine. 

I'm just wondering if we should make this change mandatory for everyone. I talked to some friends and they would prefer a clean repository containing only their latest submissions. How about making this feature optional? We could add a checkbox, like we did before for custom commit messages. What do you think?

girish-kanjiyani7 commented:
Thanks so much for reviewing the PR and for the kind words! 😊

That’s a great point — I completely understand how some users would prefer a clean repo with just the latest version.

I’d be happy to make this feature optional, similar to how the custom commit message toggle works. I can add a checkbox in the settings like:

[x] Enable Timestamped Filenames for Submissions

When enabled, filenames will include a timestamp to preserve versions. When disabled, it reverts to the original overwrite behavior.

I’ll work on the update. Let me know if you have any preferences for the checkbox label or default behavior!

raphaelheinz commented:
That would be great, thanks for your support! 

Yes, I agree with you, let's implement that way. I would label the checkbox something like "[X] Add timestamp to filenames"

Let's disable the checkbox by default so we don't break existing behavior for our plugin users. After there has been some positive feedback after some time, we could enable the checkbox by default in the future.

raphaelheinz commented:
Btw @girish-kanjiyani7, fyi yesterday I merged another PR which is a very similar use case. If you'd like you can check that out. Many thanks for your support :)

PRs: #52, #54

girish-kanjiyani7 commented:
Conflicts are resolved and the timestamp versioning feature is now working as expected (both when the toggle is on and off).

raphaelheinz commented:
That's great! Many thanks for your support. I'll merge your changes soon.

girish-kanjiyani7 commented:
I think I messed up a little by making multiple commits give me a day to make it right!

girish-kanjiyani7 commented:
Hi @raphaelheinz,
I've cleaned up the branch and removed the previous commits that were not part of the final implementation. This PR now includes only the updated and finalized version of the optional timestamp toggle feature to prevent solution file overwrite.

Please let me know if any further changes or adjustments are needed. Thanks for your time and for reviewing this!

girish-kanjiyani7 commented:
I've reviewed all the changes and it looks good to me, let me know if there's any other problem or fixes you want my help with!

raphaelheinz commented:
Hi @girish-kanjiyani7, sorry I have not merged yet and I'll finish as soon as possible. To me everything looks okay, I'm just testing if works as expected. I'll let you know.

raphaelheinz commented:
Hi @girish-kanjiyani7, finally I had some free time and reviewed. Thanks again for your work, I'm impressed and appreciate your contribution. I took this pull request as a chance for me to get more familiar with the overall code, that's why it took some time. 

In [b1271e1](https://github.com/raphaelheinz/LeetHub-3.0/pull/53/commits/b1271e14183d68e43086bc3eca27d6470b963d54) I just pushed some refactoring tasks, I think using `useTimestampFilename` as parameter was obsolete because it has not been used. Please note I've not changed anything on the functionality though. Code is just a bit simpler now.

I've one last thing to clarify before merging. Maybe you can help me with understanding: I'll create GitHub review with some notes below

raphaelheinz commented on src/js/leetcode.js:
```diff
@@ -150,8 +164,11 @@ const upload = (
       const stats = await getAndInitializeStats(problem);
-      stats.shas[problem][filename] = updatedSha;
+      const shaKey = useTimestampFilename ? filename : problem;
+      stats.shas[problem][shaKey] = updatedSha;
```
I think we should always use `const shaKey = filename;` and don't need to care if timestamps are enabled. The sha hash is unique for every single file. If we create multiple files with different timestamps, each file has different sha hash, all good. But we could also have different files when we use different programming languages (and timestamps are disabled).

So I guess we can keep the original version. Am I thinking wrong? 🤔

girish-kanjiyani7 replied:
I was originally trying to enforce a single file per problem (regardless of language) when timestamping was off, but I agree that allowing different language versions even without timestamps could be useful.

raphaelheinz commented on src/js/leetcode.js:
```diff
@@ -291,15 +308,17 @@ function uploadGit(
-          stats?.shas?.[problemName]?.[fileName] !== undefined
-            ? stats.shas[problemName][fileName]
-            : '';
+        const shaKey = useTimestampFilename ? fileName : problemName;
+        const sha = stats?.shas?.[problemName]?.[shaKey] ?? '';
```
Same here as above

girish-kanjiyani7 replied:
Yep, it's the same thing can be removed and changed to just the filename.

raphaelheinz commented on src/js/leetcode.js:
```diff
@@ -192,13 +209,12 @@ const incrementStats = () => {
-    }
-    return true;
-  });
+const checkAlreadyCompleted = async (problemName, filename) => {
+  const { useTimestampFilename = false } = await chrome.storage.local.get('useTimestampFilename');
+  const { stats } = await chrome.storage.local.get('stats');
+
+  const shaKey = useTimestampFilename ? filename : problemName;
+  return stats?.shas?.[problemName]?.[shaKey] != null;
 };
```
This method is just checking if problem has been solved before (on any programming language). So we shouldn't check on specific filename. It's sufficient to check if the problem name already existing.

girish-kanjiyani7 replied:
Thanks for the clarification! That makes complete sense — we only need to check if the problem has been solved in general, regardless of language or filename.

raphaelheinz commented:
@girish-kanjiyani7 Can you check above if everything OK from your side? After you agreed, I can push the necessary changes, test and finally merge 😃

raphaelheinz commented:
OK thanks for your feedback. I've pushed the necessary changes. I think we are ready for merge 🚀  
Many thanks for your support @girish-kanjiyani7 🙏

girish-kanjiyani7 commented:
Awesome! Thanks a ton for the review and helpful feedback throughout 🙌
Glad we could get this wrapped up.
